src/util/object_cache.h

```diff
@@ -0,0 +1,97 @@
+#ifndef CPROVER_GOTO_PROGRAMS_OBJECT_CACHE_H
+#define CPROVER_GOTO_PROGRAMS_OBJECT_CACHE_H
+
+#include <map>
+#include <memory>
+
+typedef void (*class_idt)();
+
+// This generates a static method for each template instantiation whose address
+// can be used as a unique key. We would use typeid(T) as such a key, but cbmc
+// may be built without RTTI.
+template <typename T>
+struct cached_type_idt
+{
+  static void id()
+  {
+  }
+};
+
+/// A cache of persistent objects.
+/// There can only be one instance of each type in the cache at a time.
+class object_cachet
+{
+public:
+  /// Retrives the current instance of the templated type if it exists in the cache,
+  /// creates it with the passed-in arguments and returns if not.
+  template <typename cached_type, typename... arguments>
+  cached_type &get_or_create(arguments &&... values)
+  {
+    // We're key'd off the current templates type.
+    const class_idt current_type_id = &cached_type_idt<cached_type>::id;
```

Comment:
Nit: `class_id` would match type name better than `current_type_id` (also in `erase` below).